BusinessTampereTrafficMonitoring/stream_reader/stream_reader.py

```diff
@@ -0,0 +1,102 @@
+import logging
+import queue
+import time
+from collections import deque
+from threading import Thread
+
+import cv2
+import numpy as np
+
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.ERROR)
+
+formatter = logging.Formatter('%(asctime)s:%(name)s:%message)s')
+
+file_handler = logging.FileHandler('stream_reader-log')
+file_handler.setFormatter(formatter)
+
+logger.addHandler(file_handler)
+
+
+class StreamReader:
+    def __init__(self, video_location):
+
+        self.buffer = queue.Queue()
+        self.cache = dict()
+
+        self.timestamps = deque([])
+        self.t_latest_frame = 0
+        self.video_location = video_location
+
+        self.cap = cv2.VideoCapture(self.video_location, apiPreference=cv2.CAP_FFMPEG)
+        if not self.cap.isOpened():
+            print('Cannot open stream')
+            exit(-1)
+        self.WIDTH = int(self.cap.get(cv2.CAP_PROP_FRAME_WIDTH))
+        self.HEIGHT = int(self.cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
+
+    def get_constants(self):
+        return self.WIDTH, self.HEIGHT
+
+    @staticmethod
+    def find_nearest(array, value):
+        """
+        Utility function for finding array element with closest value to value-parameter
+        """
+        array = np.asarray(array)
+        idx = (np.abs(array - value)).argmin()
+        return array[idx]
+
+    def store_frame(self, timestamp, frame):
+        """
+        Function for storing a time-stamp associated frame to cache
+        Inputs:
+            time: UNIX time stamp
+            frame: frame read using CV2, no operations done prior
+        """
+        if len(self.timestamps) == 200:
+            self.cache.pop(self.timestamps[0])
+            self.timestamps.popleft()
+        self.cache[timestamp] = frame
+        self.timestamps.append(timestamp)
+
+    def get_frame(self, epoch_time=None):
+        if epoch_time is None:
+            epoch_time = time.time()
+        return self.cache[self.find_nearest(self.timestamps, epoch_time)]
+
+    def read_stream_to_buffer(self):
+        while True:
+            success, frame = self.cap.read()
+            if success:
+                self.buffer.put(frame)
+
+    def read_buffer_to_cache(self):
+        while True:
+            t = time.time()
+            try:
+                frame = self.buffer.get()
+            except queue.Empty():
```

Comment:
Does this work? I think it should be `queue.Empty` without `()`, but I'm not 100% sure.

Reply:
Hmm, in local tests this worked but usage examples online seem to not call it either. I'll remove the brackets to be sure.